pulp_container/app/models.py

```diff
@@ -367,57 +367,52 @@ def generate_filename(instance, filename):
 
 class Upload(BaseModel):
     """
-    Model for tracking Blob uploads.
+    A model for tracking Blob uploads.
     """
 
     repository = models.ForeignKey(Repository, related_name="uploads", on_delete=models.CASCADE)
+    cumulative_size = models.BigIntegerField(default=0)
+
+
+class BlobTemporaryUpload(PulpTemporaryFile):
+    """
+    A model used for storing uploaded blob chunks in a temporary file.
+    """
 
+    upload = models.ForeignKey(Upload, related_name="blob_uploads", on_delete=models.CASCADE)
     offset = models.BigIntegerField(default=0)
 
-    file = models.FileField(
-        max_length=255,
-        null=True,
-        upload_to=generate_filename,
-        storage=FileSystemStorage(location=settings.CHUNKED_UPLOAD_DIR),
-    )
+    def save_chunk(self, chunk, chunk_size=None):
+        """Save the passed chunk to a temporary file and update offset for a current upload."""
+        self._init_temporary_file(chunk)
+        self._update_upload_size(chunk, chunk_size)
+        self._save_data()
 
-    upload_dir = os.path.join(settings.CHUNKED_UPLOAD_DIR, "container")
-
-    size = models.IntegerField(null=True)
-    md5 = models.CharField(max_length=32, null=True)
-    sha1 = models.CharField(max_length=40, null=True)
-    sha224 = models.CharField(max_length=56, null=True)
-    sha256 = models.CharField(max_length=64, null=True)
-    sha384 = models.CharField(max_length=96, null=True)
-    sha512 = models.CharField(max_length=128, null=True)
-
-    def append_chunk(self, chunk, chunk_size=None, save=True):
-        """Method for appending a chunk to a file."""
-        hashers = {}
-        for algorithm in Artifact.DIGEST_FIELDS:
-            hashers[algorithm] = getattr(hashlib, algorithm)()
-
-        self.file.close()
-        self.file.open(mode="ab")  # mode = append+binary
-        while True:
-            subchunk = chunk.read(2000000)
-            if not subchunk:
-                break
-            self.file.write(subchunk)
-            for algorithm in Artifact.DIGEST_FIELDS:
-                hashers[algorithm].update(subchunk)
+    def _init_temporary_file(self, chunk):
+        with NamedTemporaryFile("ab") as temp_file:
+            while True:
+                subchunk = chunk.read(2000000)
+                if not subchunk:
+                    break
+                temp_file.write(subchunk)
 
+            temp_file.flush()
+
+            self.file = File(open(temp_file.name, "rb"))
```

Comment:
Also this part i expect to be handled by using `PulpTemporaryFile`.

Reply:
In this case, we are dealing specifically with a chunk object which behaves differently compared to an ordinary file object. I think you need to read chunk in chunks, like so: `chunk.read(X)`. In `PulpTemporaryFile`, we are dealing just with objects of the type `File` or `PulpTemporaryUploadedFile`.

Reply:
This would require me to implement if else checks for the object's type in `PulpTemporaryFile`, would not?

Reply:
Isn't the chunk just a stream of raw data that needs to be written to a file (on whatever storage) for later use?

Maybe we should have broken this change down in two steps:
1. Handle storage without append
2. Move from local files to PTFile.

Reply:
Yes, it is.

1. The tests for S3 would not pass because I will be trying to open a file in the append mode.
2. This is what I actually did.